### PR TITLE
Run protoc as part of a makefile step optional

### PR DESCRIPTION
**Pull request by nopcoder**
In order to use `go generate` when possible - running `go generate` as part of the build and describe how to build proto files using the same technic collides. The protoc is not required in the build process as a native binary, so in this change we move the build outside the `go generate` into our Makefile.

- move protoc compile as a makefile step (not run by default)
- validator will validate protoc generated output matched the checked-in source code

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -65,36 +67,38 @@ docs-serve: ### Serve local docs
...
 	$(GOGENERATE) ./api/...
 
 del-gen-api:
 	@rm -rf $(API_BUILD_DIR)
 	@mkdir -p $(API_BUILD_DIR)
 
 .PHONY: gen-mockgen
-gen-mockgen: $(GOBINPATH)/mockgen ## Run the generator for inline commands
+gen-mockgen: go-install ## Run the generator for inline commands
 	$(GOGENERATE) ./graveler/committed/...
```
Comment (itaiad200) [inline Makefile:90]: Can we run it on the entire tree now?

**Review by itaiad200 — APPROVED** [review]

File: .github/workflows/go.yaml [hunk 2]
```diff
@@ -20,6 +20,8 @@ jobs:
         id: go
       - name: Generated code
         run: make gen
+      - name: Install Protoc
+        uses: arduino/setup-protoc@v1
```
Comment (arielshaqed) [inline .github/workflows/go.yaml:24]: arduino created this action?  No other action?? :unamused:

File: Makefile [hunk 3]
```diff
@@ -1,6 +1,7 @@
 GOCMD=$(or $(shell which go), $(error "Missing dependency - no go in PATH"))
 DOCKER=$(or $(shell which docker), $(error "Missing dependency - no docker in PATH"))
 GOBINPATH=$(shell $(GOCMD) env GOPATH)/bin
+PROTOC=$(or $(shell which protoc), $(error "Missing protobuf compilter - no protoc on PATH"))
```
Comment (arielshaqed) [inline Makefile:4]:
```suggestion
PROTOC=$(or $(shell which protoc), $(error "Missing protobuf compiler - no protoc on PATH"))
```

File: Makefile [hunk 4]
```diff
@@ -14,6 +15,7 @@ GOTEST=$(GOCMD) test
 GOTESTRACE=$(GOTEST) -race
 GOGET=$(GOCMD) get
 GOFMT=$(GOCMD)fmt
+PROTOC=protoc
```
Comment (arielshaqed) [inline Makefile:18]: Didn't you set it above?

File: Makefile (hunk 3 above)
Comment (arielshaqed) [inline Makefile:4]: Shouldn't all of these be `:=` and not `=`?  We don't need to run a shell to find which program to run before _each_ time we use one of these macros??!?

File: Makefile [hunk 5]
```diff
@@ -65,36 +67,38 @@ docs-serve: ### Serve local docs
...
 	@mkdir -p $(API_BUILD_DIR)
 
 .PHONY: gen-mockgen
-gen-mockgen: $(GOBINPATH)/mockgen ## Run the generator for inline commands
+gen-mockgen: go-install ## Run the generator for inline commands
 	$(GOGENERATE) ./graveler/committed/...
 	$(GOGENERATE) ./pyramid
 
-validate-swagger: $(GOBINPATH)/swagger  ## Validate swagger.yaml
+swagger-validator: go-install ## Validate swagger.yaml
```
Comment (arielshaqed) [inline Makefile:93]:
```suggestion
swagger-validate: go-install
```
Most of our actions are verbs in the imperative mode.  And the comment adds little.

File: Makefile [hunk 6]
```diff
@@ -130,7 +134,11 @@ fmt-validator:  ## Validate go format
 		echo Your code formatting is according to gofmt standards; \
 	fi
 
-checks-validator: lint fmt-validator validate-swagger ## Run all validation/linting steps
+.PHONY: proto-validator
+proto-validator: proto ## build proto and check if diff found
+	@git diff --quiet -- catalog/rocks/catalog.pb.go
```
Comment (arielshaqed) [inline Makefile:139]:
```suggestion
	git diff --quiet -- catalog/rocks/catalog.pb.go
```
# NOT ASHAMED TO BE RUNNING GIT!
OTOH, error messages from `git` will look strange if the last command run and displayed is `protoc`.

**Review by arielshaqed — APPROVED** [review]
Nice, thanks!

File: Makefile (hunk 5 above)
Comment (ozkatz) [reply Makefile:93]: FYI the comment is parsed and shown as part of `make help`.. not sure that's a compelling enough reason to add it though

File: Makefile (hunk 1 above)
Comment (nopcoder) [reply Makefile:90]: yes, will do it in another PR

File: .github/workflows/go.yaml (hunk 2 above)
Comment (nopcoder) [reply .github/workflows/go.yaml:24]: image
didn't find anything better in the github marketplace

File: Makefile (hunk 5 above)
Comment (nopcoder) [reply Makefile:93]: np, I'll replace all the XXX-validator to validate-XXX

File: Makefile (hunk 4 above)
Comment (nopcoder) [reply Makefile:18]: yes :)

File: Makefile (hunk 3 above)
Comment (nopcoder) [reply Makefile:4]: good point - I'll update

**Pull request merged**